dev/tool/src/db.ts

```diff
@@ -275,3 +276,94 @@ export async function moveAccountDbFromMongoToPG (
+  db: AccountDB,
+  dryRun = false
+): Promise<void> {
+  const workspaces = await listWorkspacesPure(db)
+  let updated = 0
+  for (const ws of workspaces) {
+    if (ws.uuid !== undefined) continue
+
+    const uuid = new UUID().toJSON()
+    if (!dryRun) {
+      await db.workspace.updateOne({ _id: ws._id }, { uuid })
+    }
+    updated++
+  }
+  ctx.info('Assigned uuids to workspaces', { updated, total: workspaces.length })
+}
+
+export async function updateDataWorkspaceIdToUuid (
+  ctx: MeasureMetricsContext,
+  accountDb: AccountDB,
+  dbUrl: string | undefined,
+  dryRun = false
+): Promise<void> {
+  if (dbUrl === undefined) {
+    throw new Error('dbUrl is required')
+  }
+
+  const pg = getDBClient(dbUrl)
+  try {
+    const pgClient = await pg.getClient()
+
+    // Generate uuids for all workspaces or verify they exist
+    await generateUuidMissingWorkspaces(ctx, accountDb, dryRun)
+
+    const workspaces = await listWorkspacesPure(accountDb)
+    const noUuidWss = workspaces.filter((ws) => ws.uuid === undefined)
+    if (noUuidWss.length > 0) {
+      ctx.error('Workspace uuid is required but not defined', { workspaces: noUuidWss.map((it) => it.workspace) })
+      throw new Error('workspace uuid is required but not defined')
+    }
+
+    const res = await pgClient`select t.table_name from information_schema.columns as c 
+      join information_schema.tables as t on 
+        c.table_catalog = t.table_catalog and
+        c.table_schema  = t.table_schema and 
+        c.table_name    = t.table_name
+      where t.table_type = 'BASE TABLE' and t.table_schema = 'public' and c.column_name = 'workspaceId' and c.data_type <> 'uuid'`
+    const tables: string[] = res.map((r) => r.table_name)
+
+    ctx.info('Tables to be updated: ', { tables })
+
+    for (const table of tables) {
+      ctx.info('Altering table workspaceId type to uuid', { table })
+
+      if (!dryRun) {
+        await retryTxn(pgClient, async (client) => {
+          await client`ALTER TABLE ${client(table)} RENAME COLUMN "workspaceId" TO "workspaceIdOld"`
+          await client`ALTER TABLE ${client(table)} ADD COLUMN "workspaceId" UUID`
+        })
+
+        await retryTxn(pgClient, async (client) => {
+          for (const ws of workspaces) {
+            const uuid = ws.uuid
+            if (uuid === undefined) {
+              ctx.error('Workspace uuid is required but not defined', { workspace: ws.workspace })
+              throw new Error('workspace uuid is required but not defined')
+            }
+
+            await client`UPDATE ${client(table)} SET "workspaceId" = ${uuid} WHERE "workspaceIdOld" = ${ws.workspace}`
+          }
+        })
+
+        await retryTxn(pgClient, async (client) => {
+          await client`ALTER TABLE ${client(table)} ALTER COLUMN "workspaceId" SET NOT NULL`
+        })
+
+        await retryTxn(pgClient, async (client) => {
+          await client`ALTER TABLE ${client(table)} DROP CONSTRAINT ${client(`${table}_pkey`)}`
```

Comment:
As I know this will not work for cockroach, only for PostgreSQL. Have you tested it on cockroach?

Reply:
Yes, I tested it both on Postgres and Cockroach. It works if you put it in one transaction with new PKEY creation.